### PR TITLE
Fix epoll-oneshot may cause crash (use released key)

### PR DESCRIPTION
**Pull request by jimying**
For detail , Please see Issue #3155 , thanks wahjrzd's feedback

ioqueue_add_to_set() should locked by pj_grp_lock_add_ref_dbg/pj_grp_lock_dec_ref_dbg in this case.

### CODE REVIEW

File: pjlib/src/pj/ioqueue_epoll.c [hunk 1]
```diff
@@ -819,6 +819,24 @@ PJ_DEF(int) pj_ioqueue_poll( pj_ioqueue_t *ioqueue, const pj_time_val *timeout)
...
+	 * a new event mask.
+	 */
+	pj_ioqueue_lock_key(queue[i].key);
+	if (!IS_CLOSING(queue[i].key))
+	    ioqueue_add_to_set(ioqueue, queue[i].key, 0);
+	pj_ioqueue_unlock_key(queue[i].key);
+#endif
+
 #if PJ_IOQUEUE_HAS_SAFE_UNREG
 	decrement_counter(queue[i].key);
```
Comment (jimying) [inline pjlib/src/pj/ioqueue_epoll.c:841]: line: 823   key may be released, and use it later

File: pjlib/src/pj/ioqueue_epoll.c [hunk 2]
```diff
@@ -819,6 +819,24 @@ PJ_DEF(int) pj_ioqueue_poll( pj_ioqueue_t *ioqueue, const pj_time_val *timeout)
 	    }
 	}
 
+#if USE_EPOLLONESHOT
```
Comment (sauwming) [inline pjlib/src/pj/ioqueue_epoll.c:822]: Thanks for the patch and also to @wahjrzd for the report.
But there may be a scenario when the epoll fails to be rearmed, i.e. as written in the first case of the note, when we do not wish to process a reported event (event_cnt < count). That's why initially the rearm was done outside the `event_cnt` loop (which unfortunately, caused the reported bug).

I haven't had any idea on how to fix it though.

Comment (sauwming) [reply pjlib/src/pj/ioqueue_epoll.c:822]: I had a chance to look at this again, and I think the fix looks good. But we also need to rearm the epoll in the scenario that we're not processing the event as mentioned above. Here's the additional patch (to be applied on top of this one):
[epoll_patch.txt](https://github.com/pjsip/pjproject/files/9036411/epoll_patch.txt)

Comment (jimying) [reply pjlib/src/pj/ioqueue_epoll.c:831]: Already merge the patch ,  exclude closing fd.

**Comment by sauwming** [comment]
I've merged #3151, so please merge with the latest master. Thanks.

**Comment by jimying** [comment]
> I've merged #3151, so please merge with the latest master. Thanks.

Already merged

**Review by sauwming — APPROVED** [review]

**Review by nanangizz — APPROVED** [review]

**Pull request merged**